tests/units/units.cpp

```diff
@@ -891,3 +891,164 @@ TEST(Units, complicatedMultiplicationFactorUnits)
 }
+
+TEST(Units, checkingOwningModelBothUnits)
+{
+    libcellml::ModelPtr model = libcellml::Model::create();
+
+    libcellml::UnitsPtr u = libcellml::Units::create();
+    u->setName("u");
+
+    libcellml::UnitsPtr u1 = libcellml::Units::create();
+    u1->setName("u1");
+    u1->addUnit("u", "milli", 2.0, 1000.0); //m^2
+    u1->addUnit("dimensionless");
+
+    libcellml::UnitsPtr u2 = libcellml::Units::create();
+    u2->setName("u2");
+    u2->addUnit("u", "kilo", 2.0, 0.001); // standard, exponent.
+    u2->addUnit("dimensionless");
+
+    libcellml::UnitsPtr u3 = libcellml::Units::create();
+    u3->setName("u3");
+    u3->addUnit("u", "kilo", 4.0, 0.001); // standard, exponent.
+
+    model->addUnits(u1);
+    model->addUnits(u2);
+    model->addUnits(u3);
+
+    EXPECT_EQ(0.0, libcellml::Units::scalingFactor(u1, u2));
+    EXPECT_EQ(0.0, libcellml::Units::scalingFactor(u2, u3));
+}
+
+TEST(Units, checkingOwningModelOneUnit)
+{
+    libcellml::ModelPtr model = libcellml::Model::create();
+
+    libcellml::UnitsPtr u = libcellml::Units::create();
+    u->setName("u");
+
+    libcellml::UnitsPtr u1 = libcellml::Units::create();
+    u1->setName("u1");
+    u1->addUnit("u", "milli", 1.0, 1.0); // u1 is a parent of u
+
+    libcellml::UnitsPtr u2 = libcellml::Units::create();
+    u2->setName("u2");
+    u2->addUnit("apples", "kilo", 2.0, 1000.0); // apples not in model
+
+    model->addUnits(u);
+    model->addUnits(u1);
+    model->addUnits(u2);
+
+    EXPECT_EQ(0.0, libcellml::Units::scalingFactor(u1, u2));
+    EXPECT_EQ(0.0, libcellml::Units::scalingFactor(u2, u1));
+}
+
+TEST(Units, checkScalingFactorImportedModel)
+{
+    libcellml::UnitsPtr u1 = libcellml::Units::create();
+    u1->setName("u1");
+    u1->addUnit(libcellml::Units::StandardUnit::RADIAN, 0, 1.0, 1.0);
+
+    libcellml::UnitsPtr u2 = libcellml::Units::create();
+    u2->setName("u2");
+    u2->addUnit(libcellml::Units::StandardUnit::RADIAN, 0, 1.0, 1.0);
+
+    libcellml::ModelPtr model = libcellml::Model::create();
+    model->setName("model");
+    model->addUnits(u1);
+
+    libcellml::ImportSourcePtr import = libcellml::ImportSource::create();
+    import->setUrl("I_am_a_url");
+    import->setModel(model);
+
+    model->addUnits(u1);
+    model->addUnits(u2);
+
+    EXPECT_EQ(1.0, libcellml::Units::scalingFactor(u1, u2));
+    EXPECT_EQ(1.0, libcellml::Units::scalingFactor(u2, u1));
+}
+
+TEST(Units, checkScalingFactorOneUnitImported)
+{
+    libcellml::UnitsPtr u1 = libcellml::Units::create();
+    u1->setName("u1");
+    u1->addUnit(libcellml::Units::StandardUnit::RADIAN, 0, 1.0, 1.0);
+
+    libcellml::UnitsPtr u2 = libcellml::Units::create();
+    u2->setName("u2");
+    u2->addUnit(libcellml::Units::StandardUnit::RADIAN, 0, 1.0, 1.0);
+
+    libcellml::ImportSourcePtr import = libcellml::ImportSource::create();
+    import->setUrl("I_am_a_url");
+
+    u1->setImportSource(import);
+
+    EXPECT_EQ(1.0, libcellml::Units::scalingFactor(u1, u2));
```

Comment:
I can see why this is giving `1.0` rather than `0.0` but I would expect it to fail. I suspect this is actually a bug in `Units::isBaseUnit()` which should probably be checking for imports....but need to think about this some more.

Reply:
Yeah, that would likely be it. We check if the the unit `isBaseUnit()` first, so in that test case we never actually reach the point where we would check `refUnits->isImport()`, since both `Units` are only comprised of base ones.

Reply:
I wonder if it would be even necessary to have such a condition, since the only way an imported `Unit` would fail is if there were units defined in there which were not in the original `model` and not `base`, and hence unable to be evaluated. I guess the question is if we want to ban all `imports` or just the ones which have units which are not `base`.....

Reply:
@nickerso: I've just changed `Units::updateUnitMultipliers` to return 0 for that case. I guess it would be easier to not allow any imports rather than creating a whole ton of edge cases.